src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,658 @@
+    // dio::iostream_system_err on system I/O failure.
+    bool write_nx(const char *msg) noexcept;
+    bool write(const char *msg);
+
+    bool write_nx(char msg) noexcept;
+    bool write(char msg);
+
+    bool write_nx(const std::string &msg) noexcept;
+    bool write(const std::string &msg);
+
+    bool write_nx(const int num) noexcept;
+    bool write(const int num);
+
+    bool write_nx(const unsigned num) noexcept;
+    bool write(const unsigned num);
+
+    bool write_nx(const long num) noexcept;
+    bool write(const long num);
+
+    bool write_nx(const unsigned long num) noexcept;
+    bool write(const unsigned long num);
+
+    bool write_nx(const endline &) noexcept;
+    bool write(const endline &);
+
+    bool write_nx(const flushbuf &) noexcept;
+    bool write(const flushbuf &);
+
+    template <typename A, typename B, typename ...C> inline bool write_nx(const A &a, const B &b, const C & ...c)
+            noexcept
+    {
+        if (!good()) return false;
+        if (!write_nx(a)) return false;
+        if (!write_nx(b, c...)) return false;
+        return true;
+    }
+
+    template <typename A, typename B, typename ...C> inline bool write(const A &a, const B &b, const C & ...c)
+    {
+            bool r = write_nx(a, b, c...);
+            if (!r) {
+                // This will definitely throw an exception
+                throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+            }
+            return r;
+    }
+
+    // write_buf() functions, Used to write char* or std::string.
+    //
+    // All calls to write_buf() will be buffered and not necessarily written to the file descriptor
+    // except in situations that are mentioned about write() above.
+    //
+    // Note: On partial write, io_error will reflect the cause of partial write.
+    //
+    // Returns: number of written-or-buffered characters.
+    //
+    // Non _nx variants may throw:
+    // dio::iostream_internal_err on Buffer failure (buffer did not allocated at construction time).
+    // dio::iostream_system_err on system I/O failure.
+    ssize_t write_buf_nx(const char *msg) noexcept;
+    ssize_t write_buf(const char *msg);
+
+    ssize_t write_buf_nx(const std::string &msg) noexcept;
+    ssize_t write_buf(const std::string &msg);
+
+    // This is an alias for good() function and could be used like this:
+    //
+    //      if (output_a) {
+    //          return; // Everything looks good
+    //      }
+    //
+    // Returns: true when everything looks good otherwise false.
+    explicit operator bool() noexcept;
+
+    ~ostream() noexcept;
+};
+
+/*
+ * istream
+ * =-=-=-=
+ *
+ * This class provides file input functionality. It holds a std::unique_ptr streambuf (a fixed-size
+ * circular buffer which that's size can be changed at compile-time) and controls its file descriptor
+ * and handle it gracefully when needed.
```

Comment:
This class provides a buffered input stream for file descriptors. An istream owns a file descriptor and an associated buffer.